background/index.ts

```diff
@@ -94,6 +95,103 @@ async function main() {
     { pendingTransaction: {} }
   )
 
+  // networkContext to hold the current provider and block subscriber
+  const indexBalanceContext: {
+    provider?: JsonRpcProvider
+    blockSubscriber?: (blockNumber: number) => Promise<void>
+  } = {}
+
+  const indexBalanceOnBlock = async () => {
+    // Subscriber function that handles network changes
+    const networkActiveAndAccountSubscriber = async (state: State) => {
+      console.log(
+        `[background][indexBalanceOnBlock] Network or account changed`
+      )
+
+      // Remove previous subscription if exists
+      if (indexBalanceContext.provider && indexBalanceContext.blockSubscriber) {
+        indexBalanceContext.provider.off(
+          "block",
+          indexBalanceContext.blockSubscriber
+        )
+      }
+
+      const { account, network, networkActive } = state
+
+      // Set `{ staticNetwork: true }` to avoid infinite retries if nodeRpcUrl fails.
+      // Refer: https://github.com/ethers-io/ethers.js/issues/4377
+      indexBalanceContext.provider = new JsonRpcProvider(
+        network[networkActive].nodeRpcUrl,
+        network[networkActive].chainId,
+        { staticNetwork: true }
+      )
+
+      // Define a new block subscriber
+      indexBalanceContext.blockSubscriber = async (blockNumber) => {
+        if (!network[networkActive].accountActive) {
+          return
+        }
+
+        const {
+          id,
+          tokens,
+          chainId,
+          balance: accountBalance,
+          address: accountAddress
+        } = account[network[networkActive].accountActive]
+
+        console.log(
+          `[background][indexBalanceOnBlock] Chain id: ${chainId}, New block mined: ${blockNumber}`
+        )
+
+        const nativeBalance =
+          await indexBalanceContext.provider.getBalance(accountAddress)
+
+        // Update the account balance if it has changed
+        if (number.toBigInt(accountBalance) !== nativeBalance) {
+          storage.set((state) => {
+            state.account[id].balance = number.toHex(nativeBalance)
+          })
+        }
+
+        // Update token balance if it has changed
+        tokens.forEach(async (t) => {
+          const tokenContract = await ERC20Contract.init(
+            t.address,
+            indexBalanceContext.provider
+          )
+          const tokenBalance = await tokenContract.balanceOf(accountAddress)
+
+          if (number.toBigInt(t.balance) !== tokenBalance) {
+            storage.set((state) => {
+              state.account[id].tokens.find((token) =>
+                address.isEqual(token.address, t.address)
+              ).balance = number.toHex(tokenBalance)
+            })
+          }
+        })
+      }
+
+      // Subscribe to new blocks using the updated block subscriber function
+      indexBalanceContext.provider.on(
+        "block",
+        indexBalanceContext.blockSubscriber
+      )
```

Comment:
I think maybe we can just real-time get state from storage in new block subscriber, hence we don't have to handle state sync problem.